tee-worker/litentry/core/evm-dynamic-assertions/src/precompiles/identity_to_string.rs

```diff
@@ -0,0 +1,272 @@
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::precompiles::{macros::prepare_custom_failure, PrecompileResult};
+use base58::ToBase58;
+use blake2_rfc::blake2b::Blake2b;
+use litentry_primitives::{p2pkh_address, p2sh_address, p2tr_address, p2wpkh_address};
+use ss58_registry::Ss58AddressFormat;
+use std::{format, string::String, vec, vec::Vec};
+
+pub fn identity_to_string(input: Vec<u8>) -> PrecompileResult {
+	let decoded = ethabi::decode(&[ethabi::ParamType::Uint(32), ethabi::ParamType::Bytes], &input)
+		.map_err(|e| {
+			prepare_custom_failure(format!("Could not decode input {:?}, reason: {:?}", input, e))
+		})?;
+
+	let network_type = decoded.get(0).map(|v| v.clone().into_uint().map(|t| t.as_u32())).flatten();
+	let identity_value = decoded.get(1).map(|v| v.clone().into_bytes()).flatten();
+
+	let value = match (network_type, identity_value) {
+		// Substrate
+		(Some(n), Some(v)) if n <= 6 => {
+			let network = web3_network_to_chain(n);
+			match ss58_address_of(v.as_ref(), &network) {
+				Ok(v) => Ok(v),
+				Err(e) => Err(prepare_custom_failure(format!(
+					"Cannot parse ss58 address, reason: {:?}",
+					e
+				))),
+			}
+		},
+		// Evm
+		(Some(n), Some(v)) if [7, 8, 14, 15].contains(&n) => Ok(format!("0x{}", hex::encode(&v))),
+		// Bitcoin
+		(Some(n), Some(v)) if n >= 9 && n <= 13 => {
+			let address = hex::encode(&v);
+			Ok(pubkey_to_address(n, &address))
+		},
+		// Solana
+		(Some(n), Some(v)) if n == 16 => Ok(v.to_base58()),
+		_ => Err(prepare_custom_failure(format!(
+			"Could not decode input {:?}, reason: network type or identity value is invalid",
+			input
+		))),
+	}?;
+
+	let encoded = ethabi::encode(&[ethabi::Token::String(value)]);
+	Ok(evm::executor::stack::PrecompileOutput {
+		exit_status: evm::ExitSucceed::Returned,
+		output: encoded[32..encoded.len()].to_vec(),
+	})
+}
+
+// mostly copied from https://github.com/hack-ink/substrate-minimal/blob/main/subcryptor/src/lib.rs
+// no_std version is used here
+fn ss58_address_of(public_key: &[u8], network: &str) -> core::result::Result<String, String> {
+	let network = Ss58AddressFormat::try_from(network).map_err(|e| {
+		format!("Fail to parse ss58 address, network: {:?}, reason: {:?}", network, e)
+	})?;
+	let prefix = u16::from(network);
+	let mut bytes = match prefix {
+		0..=63 => vec![prefix as u8],
+		64..=16_383 => {
+			let first = ((prefix & 0b0000_0000_1111_1100) as u8) >> 2;
+			let second = ((prefix >> 8) as u8) | ((prefix & 0b0000_0000_0000_0011) as u8) << 6;
+
+			vec![first | 0b01000000, second]
+		},
+		_ => Err(format!("Fail to parse ss58 address, network: {:?}", network))?,
+	};
+
+	bytes.extend(public_key);
+
+	let blake2b = {
+		let mut context = Blake2b::new(64);
+		context.update(b"SS58PRE");
+		context.update(&bytes);
+		context.finalize()
+	};
+
+	bytes.extend(&blake2b.as_bytes()[0..2]);
+
+	Ok(bytes.to_base58())
+}
+
+fn web3_network_to_chain(network: u32) -> String {
```

Comment:
We already have such mapping in `network_to_token`, maybe it can be reused instead ?

Reply:
network_to_token is sort of reverse of web3_network_to_chain though:
network_to_token: convert Web3Network type to numeric network code
web3_network_to_chain: convert numeric network code to Web3Network name string